### PR TITLE
Core: Replace `ip` package with `internal-ip` to address security concerns

### LINKED ISSUE #26014: [Bug]: The latest version depends on the highly vulnerable `ip` package
### Describe the bug

EDIT: New vulnerability is described here: [CVE-2024-29415](https://nvd.nist.gov/vuln/detail/CVE-2024-29415)

The vulnerability is described here: https://github.com/advisories/GHSA-78xj-cgh5-2h22.

As far as I can see now, the `ip` package is used only ones in the `core-server` package here — https://github.com/storybookjs/storybook/blob/ece1fb269cc44f43a5384f986a2b9f48613b0095/code/lib/core-server/src/utils/server-address.ts#L13C1-L13C80

This can easily be swapped with some other package, like for instance https://www.npmjs.com/package/ip-address, the way `socks`' lib maintainer did here — https://github.com/JoshGlazebrook/socks/pull/94/files.

WDYT guys? Is it a good candidate for a quick fix? I could participate if necessary with the PR, for sure.

### To Reproduce

Install any storybook flavour via `npm` (most of them depend on the vulnerable package through `core-server`)

### System

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by fyodorio**
Closes #26014

## What I did

The `ip` package has been switched out for `internal-ip` (with identical functionality coverage) because of the high-level security vulnerability in the former. The change is done at the consuming utility function and the corresponding test cases.

My research shown that the [`internal-ip` package](https://www.npmjs.com/package/internal-ip) by @sindresorhus provides sufficiently identical (to `ip.address()`) functionality via the `internalIpV4Sync()` function. So the suggestion is to go with this package further on instead of the poorly maintained (and vulnerable, see the linked issue details) `ip` package.

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This pull request has been released as version `0.0.0-pr-26025-sha-9ddb0c81`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-26025-sha-9ddb0c81 sandbox` or in an existing project with `npx storybook@0.0.0-pr-26025-sha-9ddb0c81 upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-26025-sha-9ddb0c81`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-26025-sha-9ddb0c81) |
| **Triggered by** | @valentinpalkovic |
| **Repository** | [fyodorio/storybook](https://github.com/fyodorio/storybook) |
| **Branch** | [`fyodorio/26014-fix-ip-vulnerability`](https://github.com/fyodorio/storybook/tree/fyodorio/26014-fix-ip-vulnerability) |
| **Commit** | [`9ddb0c81`](https://github.com/fyodorio/storybook/commit/9ddb0c816e343ea34ddcf3a03e9501b6c45f4d48) |
| **Datetime** | Thu Feb 15 12:53:49 UTC 2024 (`1708001629`) |
| **Workflow run** | [7916213544](https://github.com/storybookjs/storybook/actions/runs/7916213544) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=26025`_
</details>

### CODE REVIEW

**Review by izi-p — APPROVED** [review]

**Comment by valentinpalkovic** [comment]
`internal-ip` seems to be a pure ESM package. Storybook currently outputs ESM and CommonJS files. To support also CommonJS outputs we have to change the import to a dynamic import. Please be aware that top-level await isn’t supported yet AFAIK. So within the function where functions from `internal-ip` are called, try to dynamically import `import-ip` (`const importIp = await import('import-ip')`)

**Comment by fyodorio** [comment]
@valentinpalkovic I have updated the lib tests and they seem to pass successfully in CI/CD env:

Screenshot 2024-02-15 at 11 17 59

Screenshot 2024-02-15 at 11 18 44

But the suite for `cli` fails because of the similar `internal-ip` ESM import issues I guess:

Screenshot 2024-02-15 at 11 23 31

Do you have any hints on what could I do with that to make it pass?

**Comment by valentinpalkovic** [comment]
Sorry if I wasn’t precise enough. I meant to change the actual implementation instead of the test. As you can see, sandboxes cannot be generated because of the ESM issue.

File: code/lib/core-server/src/utils/server-address.ts [hunk 1]
```diff
@@ -10,7 +10,7 @@ export function getServerAddresses(
   initialPath?: string
 ) {
   const address = new URL(`${proto}://localhost:${port}/`);
-  const networkAddress = new URL(`${proto}://${host || ip.address()}:${port}/`);
+  const networkAddress = new URL(`${proto}://${host || internalIpV4Sync()}:${port}/`);
```
Comment (valentinpalkovic) [inline code/lib/core-server/src/utils/server-address.ts:13]: First, we must change `getServerAddresses` to an async function. Then some code like this should work:

```suggestion
  const internalIP = await import('internal-ip');
  const networkAddress = new URL(`${proto}://${host || internalIP.internalIpV4Sync()}:${port}/`);
```

**Comment by fyodorio** [comment]
@valentinpalkovic thanks a lot for the Midas touch, I got pretty confused tbh 👍😅

**Comment by valentinpalkovic** [comment]
Hehe. :)

I have pushed another commit. Let's see whether we get CI finally green.

**Comment by valentinpalkovic** [comment]
@fyodorio Thank you so much for your contribution so far!

**Comment by victordas** [comment]
Great to see this is fixed. Is there any ETA for this MR? 
Thanks.

**Comment by valentinpalkovic** [comment]
It’s planned for tomorrow to do some final testing, get it merged, and patched back to 7.6.

**Review by lemkaa — APPROVED** [review]

**Comment by valentinpalkovic** [comment]
During tests, I was stumbling upon an issue, where actually the internal address isn't returned.

With `ip` dependency:

image


With `internal-ip` dependency:

image


An issue is already reported: https://github.com/sindresorhus/internal-ip/issues/48, but it's from June 2023. I don't think, that `internal-ip` will fix it at any time soon.

So we have two possibilities:

a) We downgrade `internal-ip` to version 7.0.0, which is from 2021! A user reported, that the issue doesn't exist in 7.0.0
b) We use another package.

**Comment by shilman** [comment]
Fork SGTM @valentinpalkovic -- are there any other packages that are better maintained?

**Comment by cosieLq** [comment]
Maybe using [ip-address](https://www.npmjs.com/package/ip-address) like [socks](https://github.com/JoshGlazebrook/socks/releases/tag/2.7.3)?

**Comment by fyodorio** [comment]
Hi guys,

I was reviewing the `ip-address` code during work on this PR but looks like there's no directly suitable method there to swap `ip.address` with.

I wouldn't go for a fork either, just for a single method. What about just implementing that as a new utility function inside `code/lib/core-server/src/utils/server-address.ts`? It doesn't seems to be a big deal (I know, famous last words 😅). But at least it would be an independent small util further on, with a proper test(s). As this method inside `ip` didn't cause any vulnerabilities, it can be just extracted [more or less] directly (and rewritten for simplicity, as not all parts are neccessary, including the attributes).

WDYT? If you support that, I would try to come up with a suggestion during the weekend.

**Comment by cosieLq** [comment]
@fyodorio I've followed your suggestion and created a PR #26073 , just have part of the same logic as `ip.address` implemented as a small function. Not sure if it's good enough.

**Comment by fyodorio** [comment]
@cosieLq cool, looks reasonable, I'd only suggest to add the proper test coverage using one of existing functions there as an example — also if there are (should be, I believe) tests for the corresponding methods in `ip` and `internal-ip` repos, I'd reuse them to start with.

**Comment by cosieLq** [comment]
@fyodorio thank you for your suggestion. I've added a unit test to `server-address.test.ts`.

**Review by dankreiger — APPROVED** [review]

**Comment by AhmedMuhammedElsaid** [comment]
Same issue!

**Comment by valentinpalkovic** [comment]
@fyodorio 

thank you so much for your contribution and your fast reaction! 

I am closing this PR in favor of https://github.com/storybookjs/storybook/pull/2607. I guess this is also in your interest!

**Pull request closed without merging**